habitat-lab/habitat/tasks/rearrange/multi_task/pddl_defined_predicates.py

```diff
@@ -0,0 +1,384 @@
+    agent_data.grasp_mgr.update_object_to_grasp()
+
+
+def is_object_at(
+    obj: PddlEntity,
+    at_entity: PddlEntity,
+    sim_info: PddlSimInfo,
+    dist_thresh: float,
+) -> bool:
+    """
+    Checks if an object entity is logically at another entity. At an object
+    means within a threshold of that object. At a receptacle means on the
+    receptacle. At a articulated receptacle means inside of it.
+    """
+
+    entity_pos = sim_info.get_entity_pos(obj)
+
+    if sim_info.check_type_matches(
+        at_entity, SimulatorObjectType.ARTICULATED_RECEPTACLE_ENTITY.value
+    ):
+        # Object is rigid and target is receptacle, we are checking if
+        # an object is inside of a receptacle.
+        return is_inside(obj, at_entity, sim_info)
+    elif sim_info.check_type_matches(
+        at_entity, SimulatorObjectType.GOAL_ENTITY.value
+    ):
+        targ_idx = cast(
+            int,
+            sim_info.search_for_entity(at_entity),
+        )
+        idxs, pos_targs = sim_info.sim.get_targets()
+        targ_pos = pos_targs[list(idxs).index(targ_idx)]
+
+        dist = float(np.linalg.norm(entity_pos - targ_pos))
+        return dist < dist_thresh
+    elif sim_info.check_type_matches(
+        at_entity, SimulatorObjectType.STATIC_RECEPTACLE_ENTITY.value
+    ):
+        # TODO: Fix this logic to be using
+        # habitat/sims/habitat_simulator/sim_utilities.py
+        recep = cast(mn.Range3D, sim_info.search_for_entity(at_entity))
+        return recep.contains(entity_pos)
+    else:
+        raise ValueError(
+            f"Got unexpected combination of {obj} and {at_entity}"
+        )
+
+
+def set_object_at(
+    obj: PddlEntity,
+    at_entity: PddlEntity,
+    sim_info: PddlSimInfo,
+    recep_place_shrink_factor: float = 0.8,
+) -> None:
+    """
+    Sets a movable PDDL entity to match the transformation of a desired
+    `at_entity` which can be a receptacle or goal.
+
+    :param recep_place_shrink_factor: How much to shrink the size of the
+        receptacle by when placing the entity on a receptacle.
+    """
+
+    sim = sim_info.sim
+
+    # The source object must be movable.
+    if not sim_info.check_type_matches(
+        obj, SimulatorObjectType.MOVABLE_ENTITY.value
+    ):
+        raise ValueError(f"Got unexpected obj {obj}")
+
+    elif sim_info.check_type_matches(
+        at_entity, SimulatorObjectType.GOAL_ENTITY.value
+    ):
+        targ_idx = cast(
+            int,
+            sim_info.search_for_entity(at_entity),
+        )
+        all_targ_idxs, pos_targs = sim.get_targets()
+        targ_pos = pos_targs[list(all_targ_idxs).index(targ_idx)]
+        set_T = mn.Matrix4.translation(targ_pos)
+    elif sim_info.check_type_matches(
+        at_entity, SimulatorObjectType.STATIC_RECEPTACLE_ENTITY.value
+    ):
+        # Place object on top of receptacle.
+        recep = cast(mn.Range3D, sim_info.search_for_entity(at_entity))
```

Comment:
would be good to use on_top functions that we are defining so that we have a standard way of creating objects on top of receptacles, etc.

Reply:
Yes definitely. This PR is the first step towards that. I will have to target that in a future PR.